docs/architecture/014-userids.md

```diff
@@ -0,0 +1,192 @@
+application are queried, the JSON document contains internal userids
+as well as external ids:
+
+```
+GET /api/applications/1234
+{"application/id": 1234,
+ "application/events":
+ [...
+  {"event/type": "application.event/review-requested",
+   "application/id": 1234,
+   "event/actor": "5e8edd851d2fdfbd7415232c67367cc3",
+   "event/actor-attributes": {"userid": "5e8edd851d2fdfbd7415232c67367cc3",
+                              "cscId" "handler@csc.fi",
+                              "name": "Hannah Handler"}
+   "application/reviewers": [{"userid": "b58996c504c5638798eb6b511e6f49af",
+                              "eppn": "user@example.com",
+                              "name": "Example User"}],
+   ...}],
+ ...}
+```
+
+## Changing an identity
+
+Problem: A user has logged in using some identity provider and created
+some applications. However now he has started logging in via another
+identity provider and the entitlements should be associated with his
+new identity.
+
+Solution: Copy the users attributes from their new `users` row to
+their old `users` row. Remove the new `users` row. Now when the user
+logs in they get their old identity, which the entitlements are
+associated with. The external id related to those entitlements has
+also changed.
+
+# Implementation plan
+
+## 1. Add support for multiple userid attributes
+
+First off, add support for a list of `:oidc-userid-attribute` values.
+This would be an ordered list of attributes to check when the user
+logs in. The value of the first attribute that is found gets used as
+the userid.
+
+This is backwards compatible with current behaviour.
+
+[Issue #2366](https://github.com/CSCfi/rems/issues/2366),
+[PR #2471](https://github.com/CSCfi/rems/pull/2471)
+
+## 2. Workaround for THL instance
+
+The `sub` attributes are going to change when the identity provider
+for the THL instance changes. However for some time the IDP will be
+able to provide an `old_sub` attribute for the users that log in via
+the old methods (via Haka, eDuuni or Virtu).
+
+For users logging in via the new elixir method, we'll use the
+`elixirId` attribute as the user id. This will cause the right user
+ids to be used when pushing entitlements to other Elixir services.
+
+This can be accomplished by setting `:oidc-userid-attributes` to `["elixirId" "old_sub"]`.
+
+## 3. Use internal ids everywhere
+
+Implement the "Login" section of this ADR, and also write the
+necessary database migrations (see also Migration under Open questions
+below).
+
+## 4. Support for external ids in APIs
+
+Implement the "User identity resolution" section of this ADR, one API
+at a time. This should probably be in the same release as step 3.
+
+# Open questions
+
+## API format
+
+Should we switch all APIs that take userids to a structured
+`{"userid": "abc123"}` form? Then we could specify users more
+explicitly like `{"eppn": "user@example.com"}`.
+
+## `x-rems-user-id`
+
+Should we allow external user ids in `x-rems-user-id`?
```

Comment:
I think it could be nice, as long as we don't leak info in it.

Reply:
For impersonating I would think the internal id would be a bit safer option as it is much harder to guess.

Reply:
Internal ids require for the user of the API to figure out the internal id, store it somewhere and then call with it in the future. The external id is much more convenient because likely it is known without the extra steps.

I'm not sure there is any safety advantage here, the API-Key is the one that defines who can actually call the API.